packages/zoe/src/contracts/operaConcertTicket.js

```diff
@@ -0,0 +1,138 @@
+
+  // create Zoe helpers after zoe.addNewIssuer because of https://github.com/Agoric/agoric-sdk/issues/802
+  const { rejectOffer } = makeZoeHelpers(zoe);
+
+  const {
+    terms: { show, start, count },
+  } = zoe.getInstanceRecord();
+
+  const inviteHandleByTicketNumber = new Map();
+
+  function completeAmountKeywordRecord(amountKeywordRecord) {
+    const { issuerKeywordRecord } = zoe.getInstanceRecord();
+
+    const completed = { ...amountKeywordRecord };
+
+    for (const [keyword, keywordIssuer] of Object.entries(
+      issuerKeywordRecord,
+    )) {
+      if (!(keyword in completed)) {
+        completed[keyword] = keywordIssuer.getAmountMath().getEmpty();
+      }
+    }
+
+    return harden(completed);
+  }
+
+  const auditoriumSeat = harden({
+    makePaymentsAndInvites() {
+      if (inviteHandleByTicketNumber.size >= 1) {
+        throw new Error('makePaymentsAndInvites cannot be called twice');
+      }
+
+      return Array(count)
+        .fill()
+        .map((_, i) => {
+          const ticketNumber = i + 1;
+          const ticketDescription = harden({
+            show,
+            start,
+            number: ticketNumber,
+          });
+          const ticketAmount = amountMath.make(harden([ticketDescription]));
+          const payment = mint.mintPayment(ticketAmount);
+
+          const { invite, inviteHandle } = zoe.makeInvite();
+          inviteHandleByTicketNumber.set(ticketNumber, inviteHandle);
+
+          return { invite, ticketAmount, payment };
+        });
+    },
+  });
+
+  const auditoriumInvite = zoe.makeInvite(auditoriumSeat);
+
+  const makeBuyerInvite = () => {
+    const seat = harden({
+      performExchange: () => {
+        const moneyOfferHandle = inviteHandle;
+        const moneyOffer = zoe.getOffer(moneyOfferHandle);
+
+        const moneyWant = moneyOffer.proposal.want.Ticket;
+
+        const ticketNumbers = moneyWant.extent.map(t => t.number);
+        const ticketOfferHandles = ticketNumbers.map(n =>
+          inviteHandleByTicketNumber.get(n),
+        );
+
+        const offerHandles = [...ticketOfferHandles, moneyOfferHandle];
+
+        try {
+          const amountKeywordRecords = offerHandles
+            .map(offerHandle => {
+              return zoe.getOffer(offerHandle).proposal.want;
+            })
+            .map(completeAmountKeywordRecord);
+
+          zoe.reallocate(offerHandles, amountKeywordRecords);
+          zoe.complete(offerHandles);
+        } catch (err) {
+          // reallocate certainly failed
+          rejectOffer(moneyOfferHandle);
+        }
+      },
+    });
+    const { invite, inviteHandle } = zoe.makeInvite(seat);
+    return invite;
+  };
+
+  return harden({
+    invite: auditoriumInvite,
+    publicAPI: {
+      makeBuyerInvite,
+      getTicketIssuer() {
+        return issuer;
+      },
+      // This function returns a Map<TicketNumber, TicketExtent>
+      getAvailableTickets() {
+        return new Map(
```

Comment:
because of current limitation of `@agoric/marshal`, a Map is inefficient
Change this to an array of extents